src/sourcetransform.rs

```diff
@@ -304,48 +308,64 @@ impl<T> Server<T> {
     }
 
     /// Starts the gRPC server. When message is received on the `shutdown` channel, graceful shutdown of the gRPC server will be initiated.
-    pub async fn start_with_shutdown<F>(
+    pub async fn start_with_shutdown(
         &mut self,
-        shutdown: F,
+        shutdown_rx: oneshot::Receiver<()>,
     ) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
     where
         T: SourceTransformer + Send + Sync + 'static,
-        F: Future<Output = ()>,
     {
         let listener = shared::create_listener_stream(&self.sock_addr, &self.server_info_file)?;
         let handler = self.svc.take().unwrap();
-        let sourcetrf_svc = SourceTransformerService { handler };
+        let (internal_shutdown_tx, internal_shutdown_rx) = mpsc::channel(1);
+
+        let sourcetrf_svc = SourceTransformerService {
+            handler,
+            _shutdown_tx: internal_shutdown_tx,
+        };
         let sourcetrf_svc =
             proto::source_transform_server::SourceTransformServer::new(sourcetrf_svc)
                 .max_encoding_message_size(self.max_message_size)
                 .max_decoding_message_size(self.max_message_size);
 
+        let shutdown = shared::shutdown_signal(
+            internal_shutdown_rx,
+            Some(shutdown_rx),
+            CancellationToken::new(),
+        );
+
         tonic::transport::Server::builder()
             .add_service(sourcetrf_svc)
             .serve_with_incoming_shutdown(listener, shutdown)
-            .await
-            .map_err(Into::into)
+            .await?;
+
+        // cleanup the socket file after the server is shutdown
+        // UnixListener doesn't implement Drop trait, so we have to manually remove the socket file
+        let _ = fs::remove_file(&self.sock_addr);
```

Comment:
we will not remove the file if we short-circuit using `?`

Reply:
Thanks for catching this, totally missed that case.